Cgen incorrectly declares functions in output

### Describe the bug

<Title>

### Reproduction Steps

```v
import x.json2

fn maybe_map_map[T, U, X, Y](m map[T]U, f fn(T, U) !(X, Y) ) !map[X]Y {
	mut r := map[X]Y{}
	for k, v in m {
		nk, nv := f(k, v)!
		r[nk] = nv
	}
	return r
}
type Snowflake = int
struct User {
	id Snowflake
	name string
}
fn User.parse(j json2.Any) !User {
	match j {
		map[string]json2.Any {
			return User{
				id: j['id']! as int
				name: j['name']! as string
			}
		}
		else {
			return error('expected user to be object, got ${j.type_name()}')
		}
	}
}
struct Resolved {
	users ?map[Snowflake]User
}
fn Resolved.parse(j json2.Any) !Resolved {
	match j {
		map[string]json2.Any {
			return Resolved{
				users: if m := j['users'] {
					maybe_map_map[string, json2.Any, Snowflake, User](m as map[string]json2.Any, fn (k string, v json2.Any) !(Snowflake, User) {
						return Snowflake(k.int()), User.parse(v)!
					})!
				} else {
					none
				}
			}
		}
		else {
			panic('asdadsd')
		}
	}
}
fn main() {
	println('asd')
	Resolved.parse({
		'0': json2.Any('1')
		'2': '3'
		'a4': 5
		'6': '7'
	}) or {
		assert err.msg() == 'expected user to be object, got string'
		return
	}
	assert false, 'It should return error'
}
```

### Expected Behavior

Successful compile, and successful execution

### Current Behavior

```
PS D:\Games\Proekti\V\interactions> v run test.v
==================
C:/Users/mclr/AppData/Local/Temp/v_0/test.10519377724014295474.tmp.c:540: warning: WINVER redefined
C:/Users/mclr/AppData/Local/Temp/v_0/test.10519377724014295474.tmp.c:3496: error: ';' expected (got "_t5")
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Possible Solution

_No response_

### Additional Information/Context

(I'm writing more support for Interactions in [discord.v](https://github.com/DarpHome/discord.v))

### V version

V 0.4.3 c7ce72f

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.3 23ed1e9.c7ce72f
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\V\interactions
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2023-12-24 18:07:55

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: weekly.2023.51-28-gc7ce72f3
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 a39eb79b
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4911,6 +4911,7 @@ fn (mut g Gen) branch_stmt(node ast.BranchStmt) {
 }
 
 fn (mut g Gen) return_stmt(node ast.Return) {
+	g.set_current_pos_as_last_stmt_pos()
 	g.write_v_source_line_info(node.pos)
 
 	g.inside_return = true

--- a/vlib/v/tests/match_with_return_result_test.v
+++ b/vlib/v/tests/match_with_return_result_test.v
@@ -1,0 +1,68 @@
+import x.json2
+
+fn maybe_map_map[T, U, X, Y](m map[T]U, f fn (T, U) !(X, Y)) !map[X]Y {
+	mut r := map[X]Y{}
+	for k, v in m {
+		nk, nv := f(k, v)!
+		r[nk] = nv
+	}
+	return r
+}
+
+type Snowflake = int
+
+struct User {
+	id   Snowflake
+	name string
+}
+
+fn User.parse(j json2.Any) !User {
+	match j {
+		map[string]json2.Any {
+			return User{
+				id: j['id']! as int
+				name: j['name']! as string
+			}
+		}
+		else {
+			return error('expected user to be object, got ${j.type_name()}')
+		}
+	}
+}
+
+struct Resolved {
+	users ?map[Snowflake]User
+}
+
+fn Resolved.parse(j json2.Any) !Resolved {
+	match j {
+		map[string]json2.Any {
+			return Resolved{
+				users: if m := j['users'] {
+					maybe_map_map[string, json2.Any, Snowflake, User](m as map[string]json2.Any,
+						fn (k string, v json2.Any) !(Snowflake, User) {
+						return Snowflake(k.int()), User.parse(v)!
+					})!
+				} else {
+					none
+				}
+			}
+		}
+		else {
+			panic('asdadsd')
+		}
+	}
+}
+
+fn test_main() {
+	a := Resolved.parse({
+		'0':  json2.Any('1')
+		'2':  '3'
+		'a4': 5
+		'6':  '7'
+	}) or {
+		assert err.msg() == 'expected user to be object, got string'
+		return
+	}
+	assert a.users == none
+}

